### PR TITLE
Add System.Net.ServerSentEvents

### LINKED ISSUE #98105: [API Proposal]: Support parsing server-sent events (SSE)
### Background and motivation

SSE is becoming more and more popular, especially with prominent services like OpenAI relying on it for streaming responses. The format is very simple, but it still takes some amount of code to properly handle parsing the SSE format. We should have a built-in helper in either System.Net or System.Formats that take care of it for the developer (optionally then other higher-level helpers could be layered on top).

### API Proposal

```csharp
namespace System.Formats.Sse;

public readonly struct SseItem<T>
{
    public SseItem(T data, string eventType);
    public T Data { get; }
    public string EventType { get; }

    public const string EventTypeDefault = "message"; // defined as the default by SSE spec
}

public static class SseParser
{
    public static SseEnumerable<string> Parse(Stream sseStream); // equivalent to the other overload using Encoding.UTF8.GetString
    public static SseEnumerable<T> Parse<T>(Stream sseStream, SseItemParser<T> itemParser);
}

public delegate T SseItemParser<out T>(string eventType, ReadOnlySpan<byte> data);

public sealed class SseEnumerable<T> : IAsyncEnumerable<SseItem<T>>, IEnumerable<SseItem<T>>
{
    public IEnumerator<SseItem<T>> GetEnumerator();
    public IAsyncEnumerator<SseItem<T>> GetAsyncEnumerator(CancellationToken cancellationToken = default);

    public string LastEventId { get; }
    public TimeSpan ReconnectionInterval { get; }
}
```

- Ownership of the Stream is not transferred to the enumerable. Parse{Async} will not dispose of the stream, nor will the returned enumerable, and it is up to the consumer to dispose of the stream when all use of it is done.
- The enumerable will only be usable for a single enumeration. Calling Get{Async}Enumerable multiple times will fail. We could make it work in the future if really important, but for that to make sense you'd need to be working with a Stream that was seekable, and that's rare for these scenarios, which are typically networking-based.
- ReconnectionInterval defaults to Timeout.InfiniteTimeSpan. The SSE specification states that if no retry is specified in the stream, the interval may be chosen by the client. By using InfiniteTimeSpan, this conveys to a client consuming this parser that they may choose their own value.
- LastEventId defaults to empty string, per the SSE spec.

**Open Issues**
- Single type that implements both IAsyncEnumerable and IEnumerable?
- Single method that returns that single type? Naming?
- ReconnectionInterval / LastEventId on each event or on the enumerable?
- Assembly: netstandard2.0 support is needed for a variety of consumers, including the Azure SDK and Semantic Kernel. Proposal is to ship this as a new System.Formats.Sse nuget package that includes a netstandard2.0 asset. If there's an existing package it'd make sense to include this in, we could do that instead.

### API Usage

```csharp
HttpClient client = ...;
using Stream responseStream = await client.GetStreamAsync(...);
await foreach (SseItem<string> item in SseParser.Parse(responseStream, (_, bytes) => Encoding.Utf8.GetString(bytes)))
{
    Console.WriteLine(item.Data);
}
```

```csharp
HttpClient client = ...;
using Stream responseStream = await client.GetStreamAsync(...);
await foreach (SseItem<T> item in SseParser.Parse(responseStream, (_, bytes) => JsonSerializer.Deserialize<T>(bytes)))
{
    Console.WriteLine(item.Data);
}
```

### Alternative Designs

- Always returning a byte[] instead of a callback that parses a span into a T
- Also exposing a Serialize/Format method to put `SseItem<T>` instances onto a `Stream`, which ASP.NET could then use. This can come later when needed by ASP.NET.

### Risks

- We should have equivalent support in ASP.NET for serializing out an SSE stream, and it should use whatever shared `SseItem<T>` type is in the shared libraries, so we need to ensure it's designed accordingly.

### PR DESCRIPTION
**Pull request by stephentoub**
Closes https://github.com/dotnet/runtime/issues/98105

cc: @annelo-msft, @markwallace-microsoft, @trrwilson, @NTaylorMullen, @ericstj

### CODE REVIEW

File: src/libraries/System.Formats.Sse/src/System/Formats/Sse/SseEnumerable.cs [hunk 1]
```diff
@@ -0,0 +1,546 @@
...
+            {
+                // Rent a line buffer. This will grow as needed. The line buffer is what's passed to the stream,
+                // so we want it to be large enough to reduce the number of reads we need to do when data is
+                // arriving quickly. (In debug, we use a smaller buffer to stress the growth and shifting logic.)
+                _lineBuffer = ArrayPool<byte>.Shared.Rent(DefaultArrayPoolRentSize);
+                try
+                {
+                    // Spec: "Event streams in this format must always be encoded as UTF-8".
+                    // Skip a UTF8 BOM if it exists at the beginning of the stream. (The BOM is defined as optional in the SSE grammar.)
+                    while (await FillLineBufferAsync(cancellationToken).ConfigureAwait(false) != 0 && _lineLength < Utf8Bom.Length) ;
```
Comment (annelo-msft) [inline src/libraries/System.Formats.Sse/src/System/Formats/Sse/SseEnumerable.cs:198]: nit, but curious: there's a bunch of fairly detailed code shared across the sync/async versions of this method that seem to only differ by which variant of the FillLineBuffer method they call.  Wondering if there's a way to reuse these lines instead of copy?  Or would factoring to a shared method add perf overhead we seek to avoid?

Comment (stephentoub) [reply src/libraries/System.Formats.Sse/src/System/Formats/Sse/SseEnumerable.cs:198]: It's challenging to do well with the continues, breaks, and yield returns in there.

**Comment by stephentoub** [comment]
@MihaZupan, would you be able to help with a code review here?

File: src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/SseParser_1.cs [hunk 2]
```diff
@@ -0,0 +1,532 @@
...
+                _lineOffset += 3;
+                _lineLength -= 3;
+            }
+        }
+
+        /// <summary>Grows the buffer, returning the existing one to the ArrayPool and renting an ArrayPool replacement.</summary>
+        private static void GrowBuffer([NotNull] ref byte[]? buffer, int minimumLength)
+        {
+            byte[]? toReturn = buffer;
+            buffer = ArrayPool<byte>.Shared.Rent(Math.Max(minimumLength, DefaultArrayPoolRentSize));
```
Comment (MihaZupan) [inline src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/SseParser_1.cs:524]: It's a bit concerning to me that we don't have a way to limit the line length, but I guess that's a concern for threat models and docs.

**Review by MihaZupan — APPROVED** [review]
The buffer offsets are a bit hard to follow so I'd consider using `ArrayBuffer`.

Otherwise LGTM

File: src/libraries/System.Net.ServerSentEvents/src/PACKAGE.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+}
+```
+
+Synchronously parsing event contents as JSON
+
+```csharp
+MemoryStream stream = new(data);
+foreach (SseItem<Book> item in SseParser.Create(response, (eventType, bytes) => JsonSerializer.Deserialize<Book>(bytes)).Enumerate())
+{
+    Console.WriteLine(item.Author);
```
Comment (MihaZupan) [inline src/libraries/System.Net.ServerSentEvents/src/PACKAGE.md:34]:
```suggestion
    Console.WriteLine(item.Data.Author);
```

File: src/libraries/System.Net.ServerSentEvents/src/Resources/Strings.resx [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    <value>System.Resources.ResXResourceReader, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
+  </resheader>
+  <resheader name="writer">
+    <value>System.Resources.ResXResourceWriter, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
+  </resheader>
+  <data name="InvalidOperation_EnumerateOnlyOnce" xml:space="preserve">
+    <value>The enumerable may be enumerated only once.</value>
+  </data>
+  <data name="InvalidOperation_InvalidStreamNegativeBytes" xml:space="preserve">
+    <value>The Stream implementation is invalid, returning a negative number from a read operation.</value>
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/System.Net.ServerSentEvents/src/Resources/Strings.resx:124]: I'll remove it.

File: src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/SseParser_1.cs [hunk 5]
```diff
@@ -0,0 +1,532 @@
...
+            }
+            else if (fieldName.SequenceEqual("id"u8))
+            {
+                // Spec: "If the field value does not contain U+0000 NULL, then set the last event ID buffer to the field value. Otherwise, ignore the field."
+                if (fieldValue.IndexOf((byte)'\0') < 0)
+                {
+                    // Note that fieldValue might be empty, in which case LastEventId will naturally be reset to the empty string. This is per spec.
+                    LastEventId = SseParser.Utf8GetString(fieldValue);
+                }
+            }
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/SseParser_1.cs:402]: Yeah, my expectation is the id and retry in particular are pretty rare. The event name is something we might want to look at interning subsequently, if it pops as something that's frequently being sent. I don't think it's worth the complexity right now, though.

File: src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/SseParser_1.cs [hunk 6]
```diff
@@ -0,0 +1,532 @@
...
+
+        /// <summary>Indicates whether the enumerable has already been used for enumeration.</summary>
+        private int _used;
+
+        /// <summary>Buffer, either empty or rented, containing the data being read from the stream while looking for the next line.</summary>
+        private byte[] _lineBuffer = [];
+        /// <summary>The starting offset of valid data in <see cref="_lineBuffer"/>.</summary>
+        private int _lineOffset;
+        /// <summary>The length of valid data in <see cref="_lineBuffer"/>, starting from <see cref="_lineOffset"/>.</summary>
+        private int _lineLength;
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/SseParser_1.cs:53]: We could. I did it this way to make it more self-contained for now, as I'm expecting we're going to need to source-copy this to a few places until .NET 9 ships, but it's not a big deal to change it. I'll merge it as-is and we can update it subsequently.

File: src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/SseParser_1.cs [hunk 7]
```diff
@@ -0,0 +1,532 @@
...
+            _lineBuffer = ArrayPool<byte>.Shared.Rent(DefaultArrayPoolRentSize);
+            try
+            {
+                // Spec: "Event streams in this format must always be encoded as UTF-8".
+                // Skip a UTF8 BOM if it exists at the beginning of the stream. (The BOM is defined as optional in the SSE grammar.)
+                while (await FillLineBufferAsync(cancellationToken).ConfigureAwait(false) != 0 && _lineLength < Utf8Bom.Length) ;
+                SkipBomIfPresent();
+
+                // Process all events in the stream.
+                while (true)
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/SseParser_1.cs:184]: Maybe? I'm not 100% sure what you're suggesting (I think it's moving more of the logic in Enumerate{Async} into a renamed ProcessLine?), so I'll merge it as-is and then it'd be great if you could submit a PR with the recommended tweak.

**Comment by stephentoub** [comment]
Thanks for the review, @MihaZupan.
